traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go

```diff
@@ -54,6 +54,32 @@ func (cgparam *TOCacheGroupParameter) GetType() string {
 	return "cachegroup_params"
 }
 
+func (cgparam TOCacheGroupParameter) GetAuditName() string {
+	if cgparam.Name != nil {
+		return *cgparam.Name
+	}
+	if cgparam.ID != nil {
+		return strconv.Itoa(*cgparam.ID)
+	}
+	return "unknown"
+}
+
+func (cgparam TOCacheGroupParameter) GetKeyFieldsInfo() []api.KeyFieldInfo {
+	return []api.KeyFieldInfo{{"id", api.GetIntKey}}
+}
+
+func (cgparam TOCacheGroupParameter) GetKeys() (map[string]interface{}, bool) {
+	if cgparam.ID == nil {
+		return map[string]interface{}{"id": 0}, false
+	}
+	return map[string]interface{}{"id": *cgparam.ID}, true
+}
+
+func (cgparam *TOCacheGroupParameter) SetKeys(keys map[string]interface{}) {
```

Comment:
Is all of this related to Etag/IMS support for DSes and servers?

Reply:
Yes, these had to be added to make TOCacheGroupParameter fulfill the CRUDer interfaces using Modifieder.

Reply:
But the PR title seems to indicate this should only affect the `/deliveryservices` and `/servers` endpoints? Is this necessary to do that or does this also add IMS support to `/cachegroupparameters`?

Reply:
No, it doesn't add IMS support to `/cachegroupparameters`; but it prepares all CRUDer endpoints for it to be easily added in the future.

If you remove that code, the compiler gives the error
```
routing/routes.go:136:84: cannot use &cachegroupparameter.TOCacheGroupParameter literal (type *cachegroupparameter.TOCacheGroupParameter) as type api.Reader in argument to api.ReadHandler:
        *cachegroupparameter.TOCacheGroupParameter does not implement api.Reader (missing GetAuditName method)
```

Because this PR adds `Identifier` to the `api.Read` interface. That isn't strictly necessary in this PR. But we'll likely want to add unique integers to the ETag in the future, for better uniqueness than the timestamp. 

I think we should also add the unique components of the URL to the ETag, e.g. that it's a deliveryservice and its ID, to catch user errors, if the user sends the ETag without respect to the query params. For example, a user's client might automatically request `/deliveryservices`, cache the ETag, and use it again to request `/deliveryservices?42`. Which would be wrong; but the RFCs don't make that clear, they say little about query strings because it's application-specific whether the query string affects the ETag. It's an easy mistake to make, and we can easily prevent it.

Adding `Identifier` to `Reader` enables that. We could put it in `Modifieder`, but because that's currently a dangerous type assertion, it's easy to miss the Identifier and accidentally not fulfill the interface. Really, I don't see a reason why `Reader`s shouldn't be required to also be `Identifier`s. All objects should be uniquely identifiable.